fix: ensure referential integrity, match chromes behavior

DesignByOnyx opened the pull request:
Here's my initial go at it. This does the following:
- ensures that the sheet returned by `replace` and `replaceSync` point to the same sheet constructed with `new CSSStyleSheet`
- no longer overwrites the browser default `CSSStyleSheet.prototype` methods, which is always a good thing IMO.
- implements `replace` and `replaceSync` on "non-constructed" stylesheets to match Chrome's behavior.
- all tests pass

Closes #45

calebdwilliams commented on src/ConstructStyleSheet.js:
```diff
@@ -23,37 +23,16 @@ const cssStyleSheetNewMethods = ['replace', 'replaceSync'];
+           document.head.appendChild(style);
+           style.sheet.replace('body { color: blue }');
+      */
+     return Promise.reject(
```
This should only return a `Promise` for `CSSStyleSheet.prototype.replace`, `replaceSync` should only throw.

DesignByOnyx replied:
Doh, good catch.

DesignByOnyx commented on src/ConstructStyleSheet.js:
```diff
@@ -135,7 +119,43 @@ export default class ConstructStyleSheet {
+
+    actions.push([method, args]);
+
+    return result;
```
Note: this is pretty much an exact copy of the code I removed above, and this is the most significant part of the changes. Here are some notes about why I made this change:

- During init, we set [`window.CSSStyleSheet = ConstructStyleSheet`](https://github.com/calebdwilliams/construct-style-sheets/blob/master/src/index.js#L8), so I feel like the class itself should implement the full API.
- In Chrome, non-constructed stylesheets cannot be adopted. I'm not sure if this is to spec or not:
    ```js
    const style = document.createElement('style');
    document.head.appendChild(style);
    document.adoptedStyleSheets = [ style.sheet ]; // error - failed to convert value to CSSStyleSheet
    ```

calebdwilliams replied:
> In Chrome, non-constructed stylesheets cannot be adopted. I'm not sure if this is to spec or not:

Yeah, that's by design. Does your PR change that behavior?

DesignByOnyx replied:
The previous code extended the default CSSStyleSheet (referred to as OldCSSStyleSheet throughout the code) to include `replace`, `replaceSync`, and apply all changes (`addRule` and friends) to all adopted stylesheets. I feel like that behavior is wrong as it makes non-constructed stylesheets behave like constructed ones. So yes, my PR does change that.

calebdwilliams replied:
OK, just to make sure I'm not being overly-vague, I want to outline my current beliefs:

- In Chrome non-constructed stylesheets will throw if they call `replace` or `replaceSync` although oddly enough they will still perform the replacement. Not sure if that's intentional or not, maybe @rakina can help?
- Non-constructed stylesheets should not be adoptable per the spec, trying to do so should throw with `Uncaught DOMException: Failed to set the 'adoptedStyleSheets' property on 'Document': Can't adopt non-constructed stylesheets.` which is to spec so we want to throw as well.
- The `CSSStyleSheet` prototype methods should affect all adopters of that specific stylesheet.

calebdwilliams commented on src/ConstructStyleSheet.js:
```diff
@@ -83,16 +60,23 @@ export default class ConstructStyleSheet {
+  get cssRules() {
+    if (!sheetMetadataRegistry.has(this)) {
+      throw new Error(
+        "Cannot read 'cssRules' on non-constructed CSSStyleSheets.",
```
Where does this behavior come from? In Chrome you can read `cssRules`, try `document.styleSheets[0].cssRules`.

DesignByOnyx replied:
The same check is in all of the methods, so I added it here to. Basically, we throw if anybody tries to call a method or read a property directly from the prototype without first calling `new CSSStyleSheet()`. Chrome throws an `Illegal invocation` error - I can update the messaging it to match.
```
CSSStyleSheet.prototype.cssRules;
CSSStyleSheet.prototype.replace();
```

calebdwilliams replied:
Please do update the messaging. We want to be as close as possible to what Chrome does for the time being.

calebdwilliams commented on src/utils.js:
```diff
@@ -1,11 +1,5 @@
-    instance instanceof frame.CSSStyleSheet
-  );
-}
+import { instanceOfStyleSheet } from './ConstructStyleSheet';
```
No need to change this function location.

DesignByOnyx replied:
Leaving it in the old location creates a circular reference between `utils.js` and `ConstructStyleSheet.js`. It also feels more natural to have the `instanceOf` logic living close to the class it referencing (IMO). The implementation of the method might also need updating because we should not really be returning "true" for instances of the browser default class (referred to as OldCssStyleSheet throughout code).

calebdwilliams replied:
OK, I'll accept that move in that case. It's not really a problem either way, but avoiding circular dependencies is always a good idea.

We do need that check to make sure that any element passed to the adopted stylesheets array is an instance of the browser's `CSSStyleSheet`, which the `Symbol.hasInstance` takes care of for us. I'l resolve this conversation.

calebdwilliams commented on src/ConstructStyleSheet.js:
```diff
@@ -135,7 +119,43 @@ export default class ConstructStyleSheet {
+    if (!sheetMetadataRegistry.has(this)) {
+      throw new Error(
+        `Failed to execute '${method}' on 'CSSStyleSheet': Can't call ${method} on non-constructed CSSStyleSheets.`,
+      )
```
Could you give examples of what would throw this error in Chrome?

DesignByOnyx replied:
I was just copying existing logic that you already had in code. I didn't write that, just moved it.

DesignByOnyx replied:
All of these messages have been updated to "illegal invocation" messages to match Chrome.

DesignByOnyx commented on src/ConstructStyleSheet.js:
```diff
@@ -135,7 +119,43 @@ export default class ConstructStyleSheet {
+    instance instanceof OldCSSStyleSheet ||
+    (frame.CSSStyleSheet && instance instanceof frame.CSSStyleSheet)
+  );
+}
```
In reference to the discussion here: https://github.com/calebdwilliams/construct-style-sheets/pull/46#discussion_r377300040 - I think the use and implementation of this method is no longer relevant. We are effectively doing the following:

```js
class ConstructStylesSheet {
    replace() { ... }
    replaceSynce() { ... }
    addRule() { ... }
    removeRule() { ... }
    ... et al
}

window.CSSStyleSheet = ConstructStyleSheet;
```
So all checks can simply do `instanceof ConstructStyleSheet` - no need for for this method or the `Symbol.hasInstance` stuff.

calebdwilliams replied:
Sorry I've been MIA on this, but I believe the impetus for the instanceof stuff is because we still want 

```javascript
const style = document.createElement('style');
document.body.append(style);

style.sheet instanceof CSSStyleSheet;
```

and

```javascript
const sheet = new CSSStyleSheet();
sheet instanceof CSSStyleSheet
```

to both be true. Can you verify your code does that?

DesignByOnyx commented:
OK - I just pushed another set of changes as it seems like this is moving in the right direction. My second commit does the following:

- Addresses all the comments above.
- Updates all error messages to match Chrome. Everywhere we throw I included a comment with a copyable snippet which produces the error in Chrome.
- Removes the `Symbol.hasInstance` related code. I don't see why it's necessary. Please advise.
- All tests still pass.

adi518 commented:
Looks good, why it isn't merged yet?

DesignByOnyx commented:
Sorry for abandoning this - that's usually not my style. 

I have rebased my changes and updated the `instanceof` checks so that everything @calebdwilliams mentioned in [this comment](https://github.com/calebdwilliams/construct-style-sheets/pull/46#discussion_r434597729) now passes. I also updated the tests to cover those cases. All other tests still pass and this is good to go so far as I can tell.

calebdwilliams commented:
Hey no worries, 2020 has been a weird decade. Assuming everything checks out, I'll merge this and run a release later today.

calebdwilliams commented:
Looks like there are some failed tests. Probably a versioning issue. Would you mind taking a look @DesignByOnyx?

DesignByOnyx commented:
Ugh, yeah, I see it. I had just pushed a test which used `Promise.allSettled`, but the test env doesn't like that. I've fixed it.

Worth noting that somebody updated the import pattern regex a few months ago, and it was too strict. I loosened it back to what it was before... and wrote tests to prevent that from happening again (see latest commit).

I also added tests to cover all of the illegal invocation code paths, so we should have pretty good functional coverage.

calebdwilliams reviewed (approved):
Looks good to me.

calebdwilliams commented:
This PR was released as version 2.4.4
